src/coreclr/gc/gc.cpp

```diff
@@ -49307,3 +49377,294 @@ void PopulateDacVars(GcDacVars *gcDacVars)
+#endif //!MULTIPLE_HEAPS
+
+    uint32_t highmem_th_from_config = (uint32_t)GCConfig::GetGCHighMemPercent();
+    if (highmem_th_from_config)
+    {
+        high_memory_load_th = min (99, highmem_th_from_config);
+        v_high_memory_load_th = min (99, (highmem_th_from_config + 7));
+#ifdef FEATURE_EVENT_TRACE
+        high_mem_percent_from_config = highmem_th_from_config;
+#endif //FEATURE_EVENT_TRACE
+    }
+    else
+    {
+        // We should only use this if we are in the "many process" mode which really is only applicable
+        // to very powerful machines - before that's implemented, temporarily I am only enabling this for 80GB+ memory.
+        // For now I am using an estimate to calculate these numbers but this should really be obtained
+        // programmatically going forward.
+        // I am assuming 47 processes using WKS GC and 3 using SVR GC.
+        // I am assuming 3 in part due to the "very high memory load" is 97%.
+        int available_mem_th = 10;
+        if (total_physical_mem >= ((uint64_t)80 * 1024 * 1024 * 1024))
+        {
+            int adjusted_available_mem_th = 3 + (int)((float)47 / (float)g_num_processors);
+            available_mem_th = min (available_mem_th, adjusted_available_mem_th);
+        }
+
+        high_memory_load_th = 100 - available_mem_th;
+        v_high_memory_load_th = 97;
+    }
+
+    m_high_memory_load_th = min ((high_memory_load_th + 5), v_high_memory_load_th);
+}
+
+void gc_heap::refresh_memory_limit()
+{
+    if (GCConfig::GetGCTotalPhysicalMemory() != 0)
+    {
+        return;
+    }
+
+    GCToEEInterface::SuspendEE(SUSPEND_FOR_GC);
+    decommit_lock.Enter();
+
+    uint32_t nhp_from_config = static_cast<uint32_t>(GCConfig::GetHeapCount());
+#ifdef MULTIPLE_HEAPS
+    uint32_t nhp = n_heaps;
+#else
+    uint32_t nhp = 1;
+#endif //MULTIPLE_HEAPS
+    size_t seg_size_from_config;
+
+    total_physical_mem = GCToOSInterface::GetPhysicalMemoryLimit (&is_restricted_physical_mem);
+
+    bool had_hard_limit = heap_hard_limit != 0;
+#ifdef COMMITTED_BYTES_SHADOW
+    // Stress the committed bytes reconstruction logic by pretending a job object exists
+    // with memory == physical memory limit and pretending a hard limit was not there
+    // The combined result is that we will force the committed bytes reconstruction all the time.
+    is_restricted_physical_mem = true;
+    had_hard_limit = false;
+#endif //COMMITTED_BYTES_SHADOW
+
+    configure_memory_settings(false, nhp, nhp_from_config, seg_size_from_config);
+
+    if (!had_hard_limit && heap_hard_limit)
+    {
+#ifndef USE_REGIONS
+        assert (!"NYI - Segments");
+#endif //!USE_REGIONS
+        check_commit_cs.Initialize();
+
+        // Accounting for the bytes committed for the regions
+        size_t total_committed = 0;
+        size_t committed_bookkeeping = 0;
+#ifdef USE_REGIONS
+        for (int oh = soh; oh < total_oh_count; oh++)
+        {
+            int s = (oh == 0) ? 0 : oh + 2;
```

Comment:
replace 2 with max_generation. let's not use seemingly arbitrary integers. 

nit - please rename s and e to more meaning names.